File: basecode/src/main/java/ke/co/toshngure/basecode/util/BaseUtils.java
/*
 * Copyright (c) 2018.
 *
 * Anthony Ngure
 *
 * Email : anthonyngure25@gmail.com
 */

package ke.co.toshngure.basecode.util;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Point;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Build;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Display;
import android.view.Menu;
import android.view.MenuItem;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.AttrRes;
import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;


/**
 * Created by Anthony Ngure on 17/02/2017.
 * Email : anthonyngure25@gmail.com.
 */

public class BaseUtils {

    //stWyc&Y3bsb3M9

    public static float density = 1;
    public static Point displaySize = new Point();
    private static int screenHeight = 0;
    private static int screenWidth = 0;


    public static void tintProgressBar(@NonNull ProgressBar progressBar, @ColorInt int color) {
        progressBar.getIndeterminateDrawable().setColorFilter(color, PorterDuff.Mode.SRC_IN);
    }

    public static void tintImageView(ImageView imageView, @ColorInt int color) {
        imageView.setColorFilter(color);
    }







    @SuppressLint("MissingPermission")
    public static void makeCall(Context context, String phone) {
        Intent callIntent = new Intent(Intent.ACTION_CALL);
        callIntent.setData(Uri.parse("tel:" + phone));
        if (callIntent.resolveActivity(context.getPackageManager()) != null) {
            context.startActivity(Intent.createChooser(callIntent, "Make call..."));
        } else {
            Toast.makeText(context, "Unable to find a calling application.", Toast.LENGTH_SHORT).show();
        }
    }

    public static long uuidToLong(String id) {
        return (31 * 2) + ((id != null) ? id.hashCode() : 0);
    }


    public static void sendSms(Context context, String msg) {
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_TEXT, msg);
        if (intent.resolveActivity(context.getPackageManager()) != null) {
            context.startActivity(Intent.createChooser(intent, "Send sms..."));
        } else {
            Toast.makeText(context, "Unable to find a messaging application.", Toast.LENGTH_SHORT).show();
        }
    }

    public static void shareText(Context context, String text) {
        Intent sendIntent = new Intent();
        sendIntent.setAction(Intent.ACTION_SEND);
        sendIntent.putExtra(Intent.EXTRA_TEXT, text);
        sendIntent.setType("text/plain");
        Intent intent = Intent.createChooser(sendIntent, "Share with...");
        context.startActivity(intent);
    }

    public static void sendEmail(Context context, String emailAddress, String subject, String body) {
        Intent emailIntent = new Intent(Intent.ACTION_SENDTO, Uri.fromParts(
                "mailto", emailAddress, null));
        emailIntent.putExtra(Intent.EXTRA_SUBJECT, subject);
        emailIntent.putExtra(Intent.EXTRA_TEXT, body);
        context.startActivity(Intent.createChooser(emailIntent, "Send email..."));
    }


    public static void cacheInput(@NonNull EditText editText, @StringRes final int key, final PrefUtilsImpl prefUtils) {
        String currentInput = prefUtils.getString(key);
        editText.setText(currentInput);
        editText.setSelection(currentInput.length());
        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                prefUtils.writeString(key, charSequence.toString());
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });
    }

    public static int getScreenHeight(Context context) {
        if (screenHeight == 0) {
            WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
            Display display = wm.getDefaultDisplay();
            Point size = new Point();
            display.getSize(size);
            screenHeight = size.y;
        }

        return screenHeight;
    }

    public static int getScreenWidth(Context context) {
        if (screenWidth == 0) {
            WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
            Display display = wm.getDefaultDisplay();
            Point size = new Point();
            display.getSize(size);
            screenWidth = size.x;
        }

        return screenWidth;
    }


    public static int generateColor(Object object) {
        return ColorGenerator.MATERIAL.getColor(object);
    }

    public static TextDrawable getTextDrawableAvatar(String text, @Nullable Object object) {
        if (object == null) {
            object = text;
        }
        return TextDrawable.builder().round().build(text, generateColor(object));
    }

    public static int dpToPx(int dp) {
        return (int) (dp * Resources.getSystem().getDisplayMetrics().density);
    }


    /**
     * Get the color value for the given color attribute
     */
    @ColorInt
    public static int getColor(Context context, @AttrRes int colorAttrId) {
        int[] attrs = new int[]{colorAttrId /* index 0 */};
        TypedArray ta = context.obtainStyledAttributes(attrs);
        int colorFromTheme = ta.getColor(0, 0);
        ta.recycle();
        return colorFromTheme;
    }

    public static void tintMenu(Activity activity, Menu menu, @ColorInt int color) {
        if (menu != null && menu.size() > 0) {
            for (int i = 0; i < menu.size(); i++) {
                MenuItem item = menu.getItem(i);
                Drawable icon = item.getIcon();
                if (icon != null) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        icon.setTint(color);
                    } else {
                        icon.setColorFilter(color, PorterDuff.Mode.SRC_ATOP);
                    }
                }
            }
        }
    }

}

File: basecode/src/main/java/ke/co/toshngure/basecode/util/BaseUtils.kt
/*
 * Copyright (c) 2018.
 *
 * Anthony Ngure
 *
 * Email : anthonyngure25@gmail.com
 */

package ke.co.toshngure.basecode.util

import android.app.Activity
import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.graphics.Point
import android.graphics.PorterDuff
import android.net.Uri
import android.os.Build
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Patterns
import android.view.Menu
import android.view.WindowManager
import android.widget.EditText
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.Toast
import androidx.annotation.AttrRes
import androidx.annotation.ColorInt
import androidx.annotation.StringRes
import ke.co.toshngure.basecode.logging.BeeLog
import java.util.*


/**
 * Created by Anthony Ngure on 17/02/2017.
 * Email : anthonyngure25@gmail.com.
 */

object BaseUtils {

    private const val PRICE_FORMAT = "%,.2f"
    private const val CURRENCY_CODE = "KSH "
    private var screenHeight = 0
    private var screenWidth = 0

    fun isValidFullName(name: String?): Boolean {
        return name.toString().split("\\s".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray().size >= 2 && !TextUtils.isDigitsOnly(name)
    }

    fun isValidPhone(phoneNo: String?): Boolean {
        val phone = phoneNo?.trim { it <= ' ' }
        return !TextUtils.isEmpty(phone) && Patterns.PHONE.matcher(phone).matches() && phone?.length == 10
    }

    fun tintProgressBar(progressBar: ProgressBar, @ColorInt color: Int) {
        progressBar.indeterminateDrawable.setColorFilter(color, PorterDuff.Mode.SRC_IN)
    }

    fun tintImageView(imageView: ImageView, @ColorInt color: Int) {
        imageView.setColorFilter(color)
    }


    fun makeCall(context: Context, phone: String?) {
        phone?.let {
            val callIntent = Intent(Intent.ACTION_DIAL).apply {
                data = Uri.parse("tel:$phone")
            }
            try {
                if (callIntent.resolveActivity(context.packageManager) != null) {
                    context.startActivity(Intent.createChooser(callIntent, "Call with..."))
                } else {
                    Toast.makeText(context, "Unable to find a calling application.", Toast.LENGTH_SHORT).show()
                }
            } catch (e: Exception) {
                Toast.makeText(context, "Unable to make a call.", Toast.LENGTH_SHORT).show()
            }
        } ?: Toast.makeText(context, "Unable to make a call.", Toast.LENGTH_SHORT).show()

    }

    fun uuidToLong(id: String?): Long {
        return (31 * 2 + (id?.hashCode() ?: 0)).toLong()
    }

    fun formatPrice(price: Double): String {
        return CURRENCY_CODE + String.format(Locale.ENGLISH, PRICE_FORMAT, price)
    }


    fun formatPrice(price: String?): String {
        val amount = try {
            price?.toDouble() ?: 0.0
        } catch (e : Exception){
            BeeLog.e(e)
            0.0
        }
        return formatPrice(amount)
    }

    fun openUrl(context: Context, url: String) {
        val i = Intent(Intent.ACTION_VIEW)
        i.data = Uri.parse(url)
        context.startActivity(i)
    }


    fun sendSms(context: Context, phone: String?, msg: String = "") {
        phone?.let {
            val intent = Intent(Intent.ACTION_SENDTO).apply {
                type = "text/plain"
                data = Uri.parse("smsto:$phone")
                putExtra("sms_body", msg)
                putExtra(Intent.EXTRA_TEXT, msg)
            }
            try {
                if (intent.resolveActivity(context.packageManager) != null) {
                    context.startActivity(Intent.createChooser(intent, "Message with..."))
                } else {
                    Toast.makeText(context, "Unable to find a messaging application.", Toast.LENGTH_SHORT).show()
                }
            } catch (e: Exception) {
                Toast.makeText(context, "Unable to find a messaging application.", Toast.LENGTH_SHORT).show()
            }
        }
                ?: Toast.makeText(context, "Unable to find a messaging application.", Toast.LENGTH_SHORT).show()
    }

    fun shareText(context: Context, text: String) {
        val sendIntent = Intent(Intent.ACTION_SEND).apply {
            putExtra(Intent.EXTRA_TEXT, text)
            type = "text/plain"
        }
        try {
            if (sendIntent.resolveActivity(context.packageManager) != null) {
                context.startActivity(Intent.createChooser(sendIntent, "Share with..."))
            } else {
                Toast.makeText(context, "Unable to find a messaging application.", Toast.LENGTH_SHORT).show()
            }
        } catch (e: Exception) {
            Toast.makeText(context, "Unable to find a messaging application.", Toast.LENGTH_SHORT).show()
        }
    }

    fun sendEmail(context: Context, emailAddress: String, subject: String, body: String) {
        val emailIntent = Intent(Intent.ACTION_SENDTO, Uri.fromParts(
                "mailto", emailAddress, null))
        emailIntent.putExtra(Intent.EXTRA_SUBJECT, subject)
        emailIntent.putExtra(Intent.EXTRA_TEXT, body)
        context.startActivity(Intent.createChooser(emailIntent, "Send email..."))
    }


    fun cacheInput(editText: EditText, @StringRes key: Int, prefUtils: PrefUtilsImpl) {
        // Get cached text
        val currentInput = prefUtils.getString(key)
        // Set current text to the cached value
        editText.setText(currentInput)
        // Set cursor to the end of the text
        editText.setSelection(editText.text?.length ?: 0)
        editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                prefUtils.writeString(key, charSequence.toString())
            }

            override fun afterTextChanged(editable: Editable) {

            }
        })
    }

    fun getScreenHeight(context: Context): Int {
        if (screenHeight == 0) {
            val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
            val display = wm.defaultDisplay
            val size = Point()
            display.getSize(size)
            screenHeight = size.y
        }

        return screenHeight
    }

    fun getScreenWidth(context: Context): Int {
        if (screenWidth == 0) {
            val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
            val display = wm.defaultDisplay
            val size = Point()
            display.getSize(size)
            screenWidth = size.x
        }

        return screenWidth
    }


    fun dpToPx(dp: Int): Int {
        return (dp * Resources.getSystem().displayMetrics.density).toInt()
    }


    /**
     * Get the color value for the given color attribute
     */
    @ColorInt
    fun getColor(context: Context, @AttrRes colorAttrId: Int): Int {
        val attrs = intArrayOf(colorAttrId /* index 0 */)
        val ta = context.obtainStyledAttributes(attrs)
        val colorFromTheme = ta.getColor(0, 0)
        ta.recycle()
        return colorFromTheme
    }

    fun tintMenu(activity: Activity, menu: Menu?, @ColorInt color: Int) {
        if (menu != null && menu.size() > 0) {
            for (i in 0 until menu.size()) {
                val item = menu.getItem(i)
                val icon = item.icon
                if (icon != null) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        icon.setTint(color)
                    } else {
                        icon.setColorFilter(color, PorterDuff.Mode.SRC_ATOP)
                    }
                }
            }
        }
    }

    /**
     * Method checks if the app is in background or not
     */
    fun appIsInBackground(context: Context): Boolean {
        var isInBackground = true
        val am = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val runningProcesses = am.runningAppProcesses
        for (processInfo in runningProcesses) {
            if (processInfo.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND) {
                for (activeProcess in processInfo.pkgList) {
                    if (activeProcess == context.packageName) {
                        isInBackground = false
                    }
                }
            }
        }
        return isInBackground
    }


}
